crates/chain/src/tx_graph.rs

```diff
@@ -232,176 +498,392 @@ impl TxGraph {
     /// exist in `update` but not in `self`).
-    pub fn determine_additions(&self, update: &TxGraph) -> Additions {
+    pub fn determine_additions(&self, update: &TxGraph<A>) -> Additions<A> {
         let mut additions = Additions::default();
 
-        for (&txid, update_tx) in &update.txs {
-            if self.get_tx(txid).is_some() {
-                continue;
-            }
-
-            match update_tx {
-                TxNode::Whole(tx) => {
-                    if matches!(self.txs.get(&txid), None | Some(TxNode::Partial(_))) {
-                        additions.tx.insert(tx.clone());
-                    }
+        for (&txid, (update_tx_node, _, update_last_seen)) in &update.txs {
+            let prev_last_seen: u64 = match (self.txs.get(&txid), update_tx_node) {
+                (None, TxNodeInternal::Whole(update_tx)) => {
+                    additions.tx.insert(update_tx.clone());
+                    0
                 }
-                TxNode::Partial(partial) => {
-                    for (&vout, update_txout) in partial {
-                        let outpoint = OutPoint::new(txid, vout);
-
-                        if self.get_txout(outpoint) != Some(update_txout) {
-                            additions.txout.insert(outpoint, update_txout.clone());
-                        }
-                    }
+                (None, TxNodeInternal::Partial(update_txos)) => {
+                    additions.txout.extend(
+                        update_txos
+                            .iter()
+                            .map(|(&vout, txo)| (OutPoint::new(txid, vout), txo.clone())),
+                    );
+                    0
                 }
+                (Some((TxNodeInternal::Whole(_), _, last_seen)), _) => *last_seen,
+                (
+                    Some((TxNodeInternal::Partial(_), _, last_seen)),
+                    TxNodeInternal::Whole(update_tx),
+                ) => {
+                    additions.tx.insert(update_tx.clone());
+                    *last_seen
+                }
+                (
+                    Some((TxNodeInternal::Partial(txos), _, last_seen)),
+                    TxNodeInternal::Partial(update_txos),
+                ) => {
+                    additions.txout.extend(
+                        update_txos
+                            .iter()
+                            .filter(|(vout, _)| !txos.contains_key(*vout))
+                            .map(|(&vout, txo)| (OutPoint::new(txid, vout), txo.clone())),
+                    );
+                    *last_seen
+                }
+            };
+
+            if *update_last_seen > prev_last_seen {
+                additions.last_seen.insert(txid, *update_last_seen);
             }
         }
 
+        additions.anchors = update.anchors.difference(&self.anchors).cloned().collect();
+
         additions
     }
+}
+
+impl<A: Anchor> TxGraph<A> {
+    /// Get all heights that are relevant to the graph.
```

Comment:
Since this is returning a double ended iterator we have to say what the order is.

I am also concerned that this implementation is buggy. What if you alternate between `next` and `next_back` will the `last_height` state work? I don't think so. There are a two solutions I can see:

## Delete it

It seems better to wait until you have the need for this and then implement it. Its use seems speculative. We might also find it useful having an iterator of anchors that have been used. You should also be able to query "what transactions are anchored to this block id". This can all be implemented when we find the need for it.

## Simplify it

1. Make this just an `impl Iteratator`
3. Make (including document) the order from most recent to oldest. You probably either want them all or only the most recent ones.

Reply:
I simplified it.